src/codegen/events/mod.rs

```diff
@@ -0,0 +1,47 @@
+mod solana;
+mod substrate;
+
+use crate::codegen::cfg::ControlFlowGraph;
+use crate::codegen::events::solana::SolanaEventEmitter;
+use crate::codegen::events::substrate::SubstrateEventEmitter;
+use crate::codegen::vartable::Vartable;
+use crate::codegen::Options;
+use crate::sema::ast;
+use crate::sema::ast::{Function, Namespace};
+use crate::Target;
+use solang_parser::pt;
+
+/// This traits delineates the common behavior of event emission. As each target uses a different
+/// encoding scheme, there must be an implementation of this trait for each.
+pub(super) trait EventEmitter {
+    /// Generate the CFG instructions for emitting an event
```

Comment:
```suggestion
    /// Generate the CFG instructions for emitting an event.
    /// At this point, all necessary code analysis should have been done during parsing and `sema`.
    /// It is considered to be a bug if it is not possible to generate the code here.
```

Reply:
This comment is valid for everything under code generation (aka `codegen`). I don't think it is necessary to say it here, but perhaps it belongs somewhere else.

Reply:
Here maybe?
https://github.com/hyperledger/solang/blob/6134fefa43c030beae69ea8e4ba1580d3a5cc80e/src/codegen/mod.rs#L106-L111